### ISSUE TITLE
Easier derivation of Config with Scala 3

### ISSUE THREAD

**Issue by ghostdogpr** [zio/zio#9268]
There is currently no way to use the `derives` keyword to generate `Config` directly.

It is possible to do `derives DeriveConfig` but that gives you a `DeriveConfig`, not a `Config` so you still need to manually add `given Config[MyConfig] = DeriveConfig[MyConfig].desc`.

Did I miss something?

**Comment by kyri-petrou** [comment]
I think that would be nice, although it's trickier than it seems for a couple of reasons:
1. `Config` derivation (i.e., `DeriveConfig`) is not part of the core ZIO module (it's in the `zio-config-magnolia` package)
2. `Config` is sealed, so we can't extend it with a proxy class

Someone might have a better idea, but AFAICT the only way to do this without unsealing `Config` is to introduce a new mixin and use it as:

```scala
object Config {
  trait HasConfig[A] {
    def config: Config[A]
  }

  given [A](using ev: HasConfig[A]): Config[A] = ev.config
}
```

Then `DeriveConfig[A]` can extend `HasConfig[A]`. It's not pretty but I think it can get the job done. I'm also open to other ideas

**Comment by ghostdogpr** [comment]
Yeah that sounds like a correct approach given the constraints.

**Comment by kyri-petrou** [comment]
So I just thought of a different approach that doesn't require changes in ZIO, _but_ it requires an import. If we add the following in the `magnolia` package in `zio-config`:

```scala
  extension (ev: Config.type) {
    inline def derived[A](using m: Mirror.Of[A]): Config[A] = DeriveConfig.derived[A].desc
  }
```

Then the following is possible:

```scala
import zio.config.magnolia.*
import zio.*

case class Foo(value: String) derives Config
```

WDUT is better? Using a mixin or the extension method? I'm inclined towards the extension method as it requires less moving parts.

**Comment by ghostdogpr** [comment]
Yeah I agree with you, it seems lighter. And it makes sense that we need to manually import magnolia (we can imagine an alternative derivation mechanism in the future).

**Comment by jdegoes** [comment]
I think we should evaluate this ticket in light of:

> Config derivation (i.e., DeriveConfig) is not part of the core ZIO module (it's in the zio-config-magnolia package)

 - ZIO Config moving into ZIO Core.

> Config is sealed, so we can't extend it with a proxy class

 - There are some tricks we could use here if we could find a similar enough `case class` to extend. We would have to verify that it is safe if existing code is not aware of the proxy class (which seems unlikely).

> So I just thought of a different approach that doesn't require changes in ZIO, but it requires an import. If we add the following in the magnolia package in zio-config:

Not a bad idea (though of course, after the top change, could be in `zio` package itself).

**Comment by goshacodes** [comment]
Hi! Is this still relevant and what should be done? As I see - there is a way to use derives now - [#1479](https://github.com/zio/zio-config/pull/1479)

**Comment by DM-exp** [comment]
/attempt #9268
Plan: I will isolate the derivation path for Scala 3 `derives Config`, implement a minimal targeted fix in zio-config integration points, and add regression tests plus a concise note on compatibility and edge cases.

**Comment by DM-exp** [comment]
I started looking into this via /attempt #9268.

I noticed zio-config#1479 (derives `Config` with `import zio.config.magnolia.*`) was merged.

Could maintainers confirm what remains for this issue in `zio/zio` specifically (if anything), and what acceptance criteria should be for bounty payout?

Happy to proceed once scope is clarified.

**Comment by selenaalpha77-sketch** [comment]
/attempt #9268

Implemented Config.derived for Scala 3 in PR #10810. This adds the `derives` keyword support for case classes, with Scala 2 stub included.

**Comment by brone1323** [comment]
/attempt #9268

**Plan:** Add Scala 3 `derives Config` support via a `ConfigCompanionVersionSpecific` trait mixed into `object Config`, using `scala.deriving.Mirror` and `compiletime` inline machinery — no macros, no external deps, no Scala 2 changes.

Files to touch:
- `core/shared/src/main/scala/zio/Config.scala` — extend from `ConfigCompanionVersionSpecific`
- `core/shared/src/main/scala-3/zio/ConfigCompanionVersionSpecific.scala` (new) — inline `derived[A]` via Mirror + field label traversal
- `core/shared/src/main/scala-2/zio/ConfigCompanionVersionSpecific.scala` (new) — empty stub trait
- `core-tests/shared/src/test/scala-3/zio/ConfigDerivationSpec.scala` (new) — ZIOBaseSpec covering basic product, nested product, Option, List fields

This approach (approved by jdegoes + ghostdogpr in comments) adds `derived` as a method on `Config`'s companion so `case class Foo(x: String, n: Int) derives Config` works out of the box without any import from zio-config.

**Comment by iKoles-dev** [comment]
/attempt #9268

Plan: add Scala 3 `derives Config` support directly on the `Config` companion via a version-specific companion trait, keep Scala 2 behavior unchanged with an empty stub, and cover product derivation with a Scala 3 regression test for nested products, collections, and optional fields.
